docs/kf_uids/uid_considerations.md

```diff
@@ -0,0 +1,220 @@
+
+---
+
+## Key Encoding
+
+| Bits | Base32 Digits |
+|-|-|
+|64|13|
+|96|20|
+|128|26|
+|160|32|
+|192|39|
+
+| Bits | Base57 Digits |
+|-|-|
+|64|11|
+|96|17|
+|128|22|
+|160|28|
+|192|33|
+
+#### Because of how we name study buckets, we are currently constrained for studies by Amazon S3 Bucket naming rules, which stipulate:
+
+* The bucket name ... can contain only lower-case characters, numbers, periods, and dashes.
+* Each label in the bucket name must start with a lowercase letter or number.
+* The bucket name cannot contain underscores, end with a dash, have consecutive periods, or use dashes adjacent to periods.
+
+#### This means that we do one of:
+
+1. Limit ourselves to just numbers and a single lettercase for encoding.
+2. Use different ID styles for everything other than studies.
+3. Uncouple kfids from bucket names.
+
+#### If we choose 2 or 3...
+
+### Modifying Crockford's Base32 symbol set
+
+Crockford's Base32 symbol set is <font color="crimson"><ins>not perfect</ins></font> for our purpose.
+
+#### Explanation
+
+Rationales given by the design make very little sense or are self-defeating, negating any benefits from the loss of symbols.
+
+* Reducing 1/I/i/L/l to just 1 does nothing to resolve visual ambiguity for the user, because a user could just as easily read l or I instead of 1 and O instead of 0. We shouldn't demand that users internalize arbitrary technical implementation details.
+* Symbol set reductions (like from 1/I/i/L/l to 1) assume that a computer system will flexibly allow the user to enter something wrong but visually similar and then convert to the canonical symbols for uniformity. But that wouldn't help users who would look at 111 and LLL and say to themselves "but those are clearly different".
+* Capital letter L can't actually be confused with the number 1 in any font that I've ever seen, so why do we lose it?
+
+#### Possible Fix
+
+* Avoid _**both**_ parts of commonly visually ambiguous pairs, by replacing numbers "1" and "0" with capital letter "L" and "*" respectively.
+
+### Higher Bases
+
+What alphabet we use for encoding comes down to arbitrary gut feelings. https://github.com/skorokithakis/shortuuid lets you use whatever alphabet you want, which makes it a good candidate for our encoding no matter what we choose.
+
+* shortuuid defaults to using base57's "**23456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz**", which is like base58 but better, because base58 adds in "1" which we already know is commonly visually ambiguous.
+
+---
+
+## 64 bit Keys Can Provide (examples)...
+
+### Server ID + Counter
+
+* 19 bits for Server ID (500 thousand different servers)
+* 45 bits for Counter
+  * 35 trillion UIDs per server
+  * If exhausted (lol), request a new Server ID
+
+### Session ID + Counter
+
+* 44 bits for Session ID (17 trillion sessions)
```

Comment:
What is a session here?

Reply:
Like a process except if your session runs out of available keyspace before you finish then you can just ask for another one.

"Hi, I would like to give you a bunch of kfids, can I have a session ID?"
"Your session ID is 5"
"...<stream of IDs using session ID 5>..."
"Hi, I would like to give you a bunch of kfids, can I have a session ID?"
"Your session ID is 6"
"...<stream of IDs using session ID 6>..."